Move X25519/X448 to the default provider

mattcaswell opened the pull request:
This doesn't yet implement the required serializers. That will come in a subsequent PR.

paulidale commented on crypto/ec/ecx_meth.c:
```diff
@@ -1061,18 +1098,18 @@ static int s390x_pkey_ecx_keygen25519(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
     };
-    ECX_KEY *key;
+    ECX_KEY *key = NULL;
```
Setting this to NULL seems unnecessary given it is unconditionally assigned three lines later.

mattcaswell replied:
Fixed - and the other similar occurrences.

paulidale commented on crypto/ec/ecx_meth.c:
```diff
@@ -1106,18 +1141,18 @@ static int s390x_pkey_ecx_keygen448(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
     };
-    ECX_KEY *key;
+    ECX_KEY *key = NULL;
```
Same here.

paulidale commented on crypto/ec/ecx_meth.c:
```diff
@@ -1154,19 +1187,19 @@ static int s390x_pkey_ecd_keygen25519(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
     };
     unsigned char x_dst[32], buff[SHA512_DIGEST_LENGTH];
-    ECX_KEY *key;
+    ECX_KEY *key = NULL;
```
Same here.

paulidale commented on crypto/ec/ecx_meth.c:
```diff
@@ -1214,19 +1245,19 @@ static int s390x_pkey_ecd_keygen448(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
     };
     unsigned char x_dst[57], buff[114];
-    ECX_KEY *key;
+    ECX_KEY *key = NULL;
```
And here.

paulidale commented on include/crypto/ecx.h:
```diff
@@ -0,0 +1,66 @@
+
+# ifndef OPENSSL_NO_EC
+
+#include <openssl/e_os2.h>
```
Preprocessor indentation is awry.

mattcaswell replied:
Fixed.

paulidale commented on providers/implementations/exchange/ecx_exch.c:
```diff
@@ -0,0 +1,200 @@
+        return 1;
+    }
+    if (outlen < ecxctx->keylen) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
```
Would it make sense to return the desired size in this case too?  Perhaps do it unconditionally before the `secret == NULL` check?

mattcaswell replied:
I think we should avoid side effects in error conditions. I don't think its particularly useful here either.

paulidale commented on providers/implementations/keymgmt/ecx_kmgmt.c:
```diff
@@ -0,0 +1,169 @@
+        return 0;
+
+    if (key->privkey != NULL
+        && !ossl_param_bld_push_octet_string(tmpl, OSSL_PKEY_PARAM_ECX_PRIV_KEY,
```
I wonder if the param builder should be able to accept private octet and utf8 strings (storing them in secure memory)?

mattcaswell replied:
Well the param builder only stores a reference anyway so as long as the original in in secure memory then it should be fine.

paulidale commented:
Travis is also relevant.

slontis commented on crypto/ec/ecx_key.c:
```diff
@@ -0,0 +1,68 @@
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    OPENSSL_secure_free(key->privkey);
```
old code was doing secure_clear_free()

mattcaswell replied:
Fixed

slontis commented on crypto/ec/ecx_meth.c:
```diff
@@ -410,6 +397,52 @@ static int ecx_get_pub_key(const EVP_PKEY *pkey, unsigned char *pub,
+
+    ossl_param_bld_init(&tmpl);
+
+    /* A key must at least have a public part */
```
Is this always true if it generates the public part from the private part?

mattcaswell replied:
Yes. Internally if we have the private key then we have the public key too. The only time that isn't true is when the key is first created or the ECK_KEY is first constructed. But that is part of key gen or key mgmt. (Not yet moved that stuff to providers - another PR)

slontis commented on crypto/ec/ecx_meth.c:
```diff
@@ -1154,19 +1187,19 @@ static int s390x_pkey_ecd_keygen25519(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
     unsigned char x_dst[32], buff[SHA512_DIGEST_LENGTH];
-    ECX_KEY *key;
+    ECX_KEY *key = NULL;
     unsigned char *privkey = NULL, *pubkey;
```
Unless I am mistaken these 3 functions below are now identical - except for the id, and the ECErr function names.
So it could be refactored to be a single function,

mattcaswell replied:
Not quite although very similar. There are differences in digests, field multiplication function, key lengths. It could possibly still be refactored - but not in this PR.

slontis commented on include/openssl/core_names.h:
```diff
@@ -185,6 +185,10 @@ extern "C" {
 #define OSSL_PKEY_PARAM_DSA_PRIV_KEY "priv"
 
+/* ECX Keys */
+#define OSSL_PKEY_PARAM_ECX_PUB_KEY  "pub"
```
Maybe there should be a common define that is shared for these..

mattcaswell replied:
Done

slontis commented on providers/defltprov.c:
```diff
@@ -371,6 +371,8 @@ static const OSSL_ALGORITHM deflt_kdfs[] = {
 static const OSSL_ALGORITHM deflt_keyexch[] = {
 #ifndef OPENSSL_NO_DH
     { "DH:dhKeyAgreement", "default=yes", dh_keyexch_functions },
+    { "X25519", "default=yes", x25519_keyexch_functions },
```
Is there a OPENSSL_NO_ECX (or should it use NO_EC?)

mattcaswell replied:
Should be OPENSSL_NO_EC. Done

slontis commented on providers/defltprov.c:
```diff
@@ -395,6 +397,8 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
     { "DSA:dsaEncryption", "default=yes", dsa_keymgmt_functions },
 #endif
     { "RSA:rsaEncryption", "default=yes", rsa_keymgmt_functions },
+    { "X25519", "default=yes", x25519_keymgmt_functions },
```
Does this need a OPENSSL_NO_EC?

mattcaswell replied:
Yes - done.

slontis commented on providers/implementations/exchange/ecx_exch.c:
```diff
@@ -0,0 +1,200 @@
+
+/*
+ * What's passed as an actual key is defined by the KEYMGMT interface.
+ * We happen to know that our KEYMGMT simply passes DH structures, so
```
Replace DH..

mattcaswell replied:
Fixed.

slontis commented on providers/implementations/keymgmt/ecx_kmgmt.c:
```diff
@@ -0,0 +1,169 @@
+    pubkey = key->pubkey;
+    if ((param_priv_key != NULL
+             && !OSSL_PARAM_get_octet_string(param_priv_key,
+                                            (void **)&key->privkey, keylen,
```
Not your fault, but I shudder everytime I see this (void **)& in OSSL_PARAM_get_octet_string

mattcaswell replied:
Yes - I agree.

slontis commented on providers/implementations/keymgmt/ecx_kmgmt.c:
```diff
@@ -0,0 +1,169 @@
+             && !OSSL_PARAM_get_octet_string(param_priv_key,
+                                            (void **)&key->privkey, keylen,
+                                             &privkeylen))
+            || !OSSL_PARAM_get_octet_string(param_pub_key,
```
Hmm this might be more readable if there are 2 if's here..
One that just handles the  private bit and does the privlen check
And one that does the public bit and public length check.

mattcaswell replied:
Done

romen commented:
Just cross-mentioning the related issues: #10353 and #10354

mattcaswell commented:
Fixup commits pushed addressing the above comments and the travis failure.

Please take another look.

mattcaswell commented:
Oh...and also I realised I missed off the s390 support. Now added.

@p-steuer

slontis commented on crypto/s390xcpuid.S:
```diff
@@ -0,0 +1,428 @@
+#include "s390x_arch.h"
+
+.text
+
```
Where has this file come from?

p-steuer replied:
Its generated by s390xcpuid.pl, @mattcaswell  probably added it accidentially.

mattcaswell replied:
Yes I did. I'll remove it.

mattcaswell commented:
Fixed the s390xcpuid.S issue. The travis red cross was due to a bad auto-merge with master during the travis build. I've rebased to resolve that issue.

Please take another look.

p-steuer commented:
With moving the s390 code to provider, the corresponding run-time feature detection was lost. My tests of the S390X_EC_ASM build pass on z15 but it should SIGILL on any older machine, since ECC instruction set extension is not available there.

Specifically im talking about the following (non provider) code from crypto/ec/ecx_meth.c:
```
const EVP_PKEY_METHOD *ecx25519_pkey_method(void)
{
#ifdef S390X_EC_ASM
    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X25519))
        return &ecx25519_s390x_pkey_meth;
#endif
    return &ecx25519_pkey_meth;
}

const EVP_PKEY_METHOD *ecx448_pkey_method(void)
{
#ifdef S390X_EC_ASM
    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X448))
        return &ecx448_s390x_pkey_meth;
#endif
    return &ecx448_pkey_meth;
}
```

This could be easily fixed by checking the above conditions inside provider code's S390X_EC_ASM ifdefs (before every operation).

However, since this check is only required one time at init, i would prefer a solution similar to the non-provider code (to also get rid of a lot of ifdefs and to facilitate integration of future platform specific hw support). E.g. replace member pointing to an OPENSSL_DISPATCH object in OPENSSL_ALGORITHM structure by a member pointing to a function returning such objects. That way, you could define both a platform-spefific and a generic OPENSSL_DISPATCH array, and said function could chose which one to return, based on the hw capabilities detected on init (analogous to what i did to the standard_methods array here: 19bd1fa1ef17ff0ac071402cb6b002f1c29e37b9).

mattcaswell commented:
> This could be easily fixed by checking the above conditions inside provider code's S390X_EC_ASM ifdefs (before every operation).

Fixed. I have gone with this simple solution for now. I did look at an option for a one time capability check. We actually have a mechanism for that for ciphers. See:

https://github.com/openssl/openssl/blob/7fa8bcfe4342df41919f5564b315f9c85d0a02d6/providers/defltprov.c#L190-L197

It would not take much to extend that to key exchange. However in this case I think the benefit is very minimal at the cost of quite a lot of extra plumbing. I'm not actually sure you would gain anything very much.

p-steuer commented on providers/implementations/exchange/ecx_exch.c:
```diff
@@ -0,0 +1,223 @@
+    if (ecxctx->keylen == X25519_KEYLEN) {
+#ifdef S390X_EC_ASM
+        if (OPENSSL_s390xcap_P.pcc[1]
+                & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X448)) {
```
S390X_SCALAR_MULTIPLY_X448 -> S390X_SCALAR_MULTIPLY_X25519

mattcaswell replied:
Oops. Fixed.

slontis commented:
needs a rebase

slontis commented:
@mattcaswell

mattcaswell commented:
> needs a rebase

Sigh. Needs a bit more work due to the all the keymgmt changes in master.

mattcaswell commented:
I rebased this and reworked it to take account of the refactored keymgmt in master.

Please take another look and reconfirm?

p-steuer commented on crypto/ec/ecx_meth.c:
```diff
@@ -416,6 +403,48 @@ static int ecx_get_pub_key(const EVP_PKEY *pkey, unsigned char *pub,
+
+ err:
+    ossl_param_bld_free(params);
+    return rv;
```
rv may be uninitialized here

mattcaswell replied:
Fixed

p-steuer commented on providers/implementations/keymgmt/ecx_kmgmt.c:
```diff
@@ -0,0 +1,231 @@
+        return 0;
+
+    if (param_priv_key != NULL
+             && !OSSL_PARAM_get_octet_string(param_priv_key,
```
param_priv_key maybe uninitialized here. Should probably be initialized to NULL.

mattcaswell replied:
Fixed.

mattcaswell commented:
Fixups pushed addressing both of those issues.

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Pushed. Thanks.
